File: internal/services/user_service.go
package services

import (
	"pocket-app/internal/config"
	"pocket-app/pkg/logger"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
)

// UserService handles user-related business logic
type UserService struct {
	app    *pocketbase.PocketBase
	config *config.Config
}

// NewUserService creates a new user service
func NewUserService(app *pocketbase.PocketBase, cfg *config.Config) *UserService {
	return &UserService{
		app:    app,
		config: cfg,
	}
}

// GetUserProfile retrieves a user profile with extended information
func (s *UserService) GetUserProfile(userID string) (*core.Record, error) {
	// Get user record
	user, err := s.app.FindRecordById("_pb_users_auth_", userID)
	if err != nil {
		logger.Error("Failed to find user", err)
		return nil, err
	}

	// Try to get extended profile
	profile, err := s.app.FindFirstRecordByFilter(
		"user_profiles", 
		"user = {:id}",
		map[string]interface{}{"id": userID},
	)
	if err != nil {
		logger.Debug("No extended profile found for user", userID)
		// Return basic user info even if extended profile doesn't exist
		return user, nil
	}

	// You can enhance this to merge user and profile data
	return profile, nil
}

// CreateUserProfile creates an extended user profile
func (s *UserService) CreateUserProfile(userID string, data map[string]interface{}) (*core.Record, error) {
	collection, err := s.app.FindCollectionByNameOrId("user_profiles")
	if err != nil {
		logger.Error("Failed to find user_profiles collection", err)
		return nil, err
	}

	profile := core.NewRecord(collection)
	profile.Set("user", userID)
	
	// Set profile data
	for key, value := range data {
		if key != "user" { // Prevent overriding user field
			profile.Set(key, value)
		}
	}

	if err := s.app.Save(profile); err != nil {
		logger.Error("Failed to create user profile", err)
		return nil, err
	}

	return profile, nil
}

// UpdateUserProfile updates an existing user profile
func (s *UserService) UpdateUserProfile(userID string, data map[string]interface{}) (*core.Record, error) {
	profile, err := s.app.FindFirstRecordByFilter(
		"user_profiles", 
		"user = {:id}",
		map[string]interface{}{"id": userID},
	)
	if err != nil {
		logger.Error("Failed to find user profile", err)
		return nil, err
	}

	// Update profile data
	for key, value := range data {
		if key != "user" { // Prevent overriding user field
			profile.Set(key, value)
		}
	}

	if err := s.app.Save(profile); err != nil {
		logger.Error("Failed to update user profile", err)
		return nil, err
	}

	return profile, nil
}

// GetPublicProfiles retrieves public user profiles with pagination
func (s *UserService) GetPublicProfiles(page, perPage int, search string) ([]*core.Record, error) {
	filter := "is_public = true"
	params := map[string]interface{}{}
	
	if search != "" {
		filter += " && (display_name ~ {:search} || user.username ~ {:search})"
		params["search"] = search
	}

	records, err := s.app.FindRecordsByFilter(
		"user_profiles",
		filter,
		"-created",
		perPage,
		(page-1)*perPage,
	)
	if err != nil {
		logger.Error("Failed to get public profiles", err)
		return nil, err
	}

	return records, nil
}

File: backend/src/services/user-service.ts
/**
 * User Service - TypeScript implementation
 * Migrated from Go service with enhanced Rivet integration
 */

/// <reference path="../../../pb_data/types.d.ts" />

import { Logger } from '../utils/logger';

export class UserService {
  private logger: Logger;

  constructor() {
    this.logger = new Logger('UserService');
  }

  /**
   * Get user profile with extended information
   */
  async getUserProfile(userID: string): Promise<models.Record | null> {
    try {
      // Get user record
      const user = $app.dao().findRecordById('_pb_users_auth_', userID);
      if (!user) {
        this.logger.error('User not found', { userID });
        return null;
      }

      // Try to get extended profile
      try {
        const profile = $app.dao().findFirstRecordByFilter(
          'user_profiles',
          'user = {:id}',
          { id: userID }
        );
        
        this.logger.debug('Extended profile found', { userID });
        return profile;
      } catch (error) {
        this.logger.debug('No extended profile found, returning basic user', { userID });
        return user;
      }
    } catch (error) {
      this.logger.error('Failed to get user profile', error);
      return null;
    }
  }

  /**
   * Create extended user profile
   */
  async createUserProfile(userID: string, data: Record<string, any>): Promise<models.Record | null> {
    try {
      const collection = $app.dao().findCollectionByNameOrId('user_profiles');
      if (!collection) {
        throw new Error('user_profiles collection not found');
      }

      const profile = new Record(collection);
      profile.set('user', userID);
      
      // Set profile data
      for (const [key, value] of Object.entries(data)) {
        if (key !== 'user') { // Prevent overriding user field
          profile.set(key, value);
        }
      }

      $app.dao().saveRecord(profile);
      this.logger.info('User profile created', { userID, profileId: profile.id });
      
      return profile;
    } catch (error) {
      this.logger.error('Failed to create user profile', error);
      return null;
    }
  }

  /**
   * Update existing user profile
   */
  async updateUserProfile(userID: string, data: Record<string, any>): Promise<models.Record | null> {
    try {
      const profile = $app.dao().findFirstRecordByFilter(
        'user_profiles',
        'user = {:id}',
        { id: userID }
      );

      if (!profile) {
        this.logger.error('User profile not found', { userID });
        return null;
      }

      // Update profile data
      for (const [key, value] of Object.entries(data)) {
        if (key !== 'user') { // Prevent overriding user field
          profile.set(key, value);
        }
      }

      $app.dao().saveRecord(profile);
      this.logger.info('User profile updated', { userID, profileId: profile.id });
      
      return profile;
    } catch (error) {
      this.logger.error('Failed to update user profile', error);
      return null;
    }
  }

  /**
   * Get public user profiles with pagination
   */
  async getPublicProfiles(page: number = 1, perPage: number = 20, search: string = ''): Promise<models.Record[]> {
    try {
      let filter = 'is_public = true';
      const params: Record<string, any> = {};
      
      if (search) {
        filter += ' && (display_name ~ {:search} || user.username ~ {:search})';
        params.search = search;
      }

      const records = $app.dao().findRecordsByFilter(
        'user_profiles',
        filter,
        '-created',
        perPage,
        (page - 1) * perPage,
        params
      );

      this.logger.debug('Public profiles retrieved', { count: records.length, page, search });
      return records;
    } catch (error) {
      this.logger.error('Failed to get public profiles', error);
      return [];
    }
  }

  /**
   * Validate user permissions for operation
   */
  validateUserPermissions(userId: string, operation: string, resourceId?: string): boolean {
    try {
      // Basic permission logic - can be enhanced based on your needs
      const user = $app.dao().findRecordById('_pb_users_auth_', userId);
      if (!user) return false;

      // Add your permission logic here
      // For example, check user roles, ownership, etc.
      
      this.logger.debug('User permissions validated', { userId, operation, resourceId });
      return true;
    } catch (error) {
      this.logger.error('Failed to validate user permissions', error);
      return false;
    }
  }

  /**
   * Get user statistics
   */
  async getUserStats(userId: string): Promise<Record<string, any> | null> {
    try {
      const user = $app.dao().findRecordById('_pb_users_auth_', userId);
      if (!user) return null;

      // Count user's posts
      const postsCount = $app.dao().findRecordsByFilter(
        'posts',
        'author = {:userId}',
        '',
        1,
        0,
        { userId }
      ).length;

      // Count user's comments
      const commentsCount = $app.dao().findRecordsByFilter(
        'comments',
        'author = {:userId}',
        '',
        1,
        0,
        { userId }
      ).length;

      const stats = {
        postsCount,
        commentsCount,
        joinedAt: user.created,
        lastActive: user.updated
      };

      this.logger.debug('User stats retrieved', { userId, stats });
      return stats;
    } catch (error) {
      this.logger.error('Failed to get user stats', error);
      return null;
    }
  }
}
